### DISCUSSION TITLE
Discriminate union based on Model key (instead of field value)

### DISCUSSION THREAD

**Discussion by winpat** [pydantic/pydantic#8002] [General]
I have the following JSON structure which I would like to parse:
``` json
[
    {"cat": {"age": 1}},
    {"dog": {"age": 1}}
]
```
And the following models:
``` python
class Cat(BaseModel):
    age: int

class Dog(BaseModel):
    age: int
```

According to @samuelcolvin in #4180 this should be possible in Pydantic v2, but I could not find any examples on how to do this without falling back to the smart union behavior.

Is there a way to discriminate based on "cat" or "dog"?

**Comment by sydney-runkle** [comment]
Great question! Yes - we just introduced a `Discriminator` object that can be used with either a field name (as you're used to) or a callable. This feature will be released in v2.5: https://docs.pydantic.dev/dev/concepts/unions/#discriminated-unions-with-callable-discriminators
